README.md
# Vishnu

[![Gem](https://img.shields.io/gem/v/vishnu.svg?style=flat-square)](https://rubygems.org/gems/commonmarker-rouge)
[![Gitlab pipeline status](https://img.shields.io/gitlab/pipeline/sandfox/vishnu/master.svg?style=flat-square)](https://gitlab.com/sandfox/vishnu/-/pipelines)

Vishnu is a simple library to use Libravatar avatars in your ruby app.

[Libravatar](https://libravatar.org/) is an avatar service to let their
users associate avatar images with their emails or openids. This rubygem
generates their avatar URL.

## Installation

Add the following line to your ```Gemfile```:

```ruby
gem 'vishnu'
```

Or if you want to register ```Libravatar``` alias, then:

```ruby
gem 'vishnu', require: 'libravatar'
```

## Simple Usage

```ruby
Vishnu.new(email:  'someone@example.com').url   # get avatar for email
Vishnu.new(openid: 'https://example.com').url   # get avatar for OpenID URL
```

## Documentation

Read full documentation here: <https://sandfox.dev/ruby/vishnu.html>

## Support

Please file issues on our main repo at GitLab: <https://gitlab.com/sandfox/vishnu/-/issues>

## License

Licensed under the MIT License. See ```LICENSE.txt``` for further details.


CHANGELOG.md
# Changelog

## 2.x

### 2.0.1

*Dec 9, 2023*

* Re-release 2.0.0 with updated homepage and email

### 2.0.0

*26 Mar, 2016*

Forked from 1.2.0 of the Libravatar gem

Mostly refactoring
- use `url` method as primary way to get avatar url instead of `to_s`
- make all methods except for `url`, `to_s` and attribute setters and getters private
- bump minimum ruby version to 2.0.0
- change build system from jeweler to bundler
- change test system from shoulda to rspec
- add travis builds for all minor ruby versions

## 1.x

### 1.2.2

*20 Jun, 2016*

Merged development dependencies from upstream

### 1.2.1

*25 Mar, 2016*

bugfix release for libravatar 1.2.0
- fix weighted random on federated avatars
- fix crash on custom ports on federated avatars

UPD: contains all fixes from libravatar 1.3.0


spec/spec_helper.rb
if ENV['COVERAGE'] == '1'
  require 'simplecov'
  SimpleCov.start
end

require_relative '../lib/vishnu'


spec/vishnu_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe Vishnu do
  describe 'Gets urls' do
    before(:each) do
      dns = double('Resolv::DNS')
      allow(Resolv::DNS).to receive(:open).and_yield(dns)
      allow(dns).to receive(:getresources) do |host, type|
        expect(type).to eq(Resolv::DNS::Resource::IN::SRV)

        case host.downcase
        when '_avatars._tcp.federated.com'
          [
            Resolv::DNS::Resource::IN::SRV.new(
              0, 5, 80, 'avatars.federated.com'
            ),
            Resolv::DNS::Resource::IN::SRV.new(
              5, 5, 80, 'not-in-priority.federated.com'
            ),
          ]
        when '_avatars-sec._tcp.federated.com'
          [
            Resolv::DNS::Resource::IN::SRV.new(
              5, 5, 443, 'not-in-priority.federated.com'
            ),
            Resolv::DNS::Resource::IN::SRV.new(
              0, 5, 443, 'avatars.federated.com'
            ),
          ]
        when '_avatars._tcp.custom-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 8080, 'avatars.custom-federated.com'
          )]
        when '_avatars-sec._tcp.custom-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 8043, 'avatars.custom-federated.com'
          )]
        when '_avatars._tcp.nonjunk-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 12345, 'hosntame.abcde.fghi.com'
          )]
        when '_avatars._tcp.junk-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 65348283, 'FNORD IMPUNTK *#(*$#&'
          )]
        when '_avatars._tcp.junk2-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 65348283, 'hosntame.abcde.fghi.com'
          )]
        when '_avatars._tcp.junk3-federated.com'
          [Resolv::DNS::Resource::IN::SRV.new(
            0, 5, 12345, 'FNORD IMPUNTK *#(*$#&'
          )]
        else
          []
        end
      end
    end

    it 'gets url for non federated email' do
      # non-federated
      avatar = Vishnu.new(email: 'user@example.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')

      avatar = Vishnu.new(email: 'user@example.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')

      avatar = Vishnu.new(email: 'USER@ExAmPlE.CoM')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')

      avatar = Vishnu.new(email: 'user@example.com', https: true)
      expect(avatar.url).to eq('https://seccdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')

      avatar = Vishnu.new(email: 'user@example.com', https: false)
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')

      avatar = Vishnu.new(email: 'USER@ExAmPlE.CoM', default: 'http://example.com/avatar.png')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af?d=http://example.com/avatar.png')

      avatar = Vishnu.new(email: 'USER@ExAmPlE.CoM', size: 512, default: 'mm')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af?s=512&d=mm')
    end

    it 'gets url for federated email' do
      # federated
      avatar = Vishnu.new(email: 'user@federated.com')
      expect(avatar.url).to eq('http://avatars.federated.com/avatar/d69b469ded547b3ddef720a70c186322')

      # federated secure
      avatar = Vishnu.new(email: 'user@feDeRaTed.cOm', https: true)
      expect(avatar.url).to eq('https://avatars.federated.com/avatar/d69b469ded547b3ddef720a70c186322')

      # federated with custom port
      avatar = Vishnu.new(email: 'USER@cuStOm-feDerated.COM')
      expect(avatar.url).to eq('http://avatars.custom-federated.com:8080/avatar/8df8704e4b556e0684f7c38accdaf517')

      # federated secure with custom port
      avatar = Vishnu.new(email: 'user@custom-federated.com', https: true)
      expect(avatar.url).to eq('https://avatars.custom-federated.com:8043/avatar/8df8704e4b556e0684f7c38accdaf517')
    end

    it 'gets url for non federated openid' do
      avatar = Vishnu.new(openid: 'http://example.com/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd')

      avatar = Vishnu.new(openid: 'https://example.com/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/82be86d8b10f6492d0eb3d6475c388044529b9d4ddf7269dec2483601b22d2e1')

      avatar = Vishnu.new(openid: 'hTTp://EXAMPLE.COM/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd')

      avatar = Vishnu.new(openid: 'hTTp://EXAMPLE.COM/id/Bob', size: 512)
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd?s=512')

      avatar = Vishnu.new(openid: 'http://example.com/id/bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/84106b275be9a7c69f3b0f77d1d504a794e6d0e4e0a068fa529d869b721f4261')

      avatar = Vishnu.new(openid: 'hTTp://EXAMPLE.COM/ID/BOB')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/10e678f0db6ead293f21fae8adb1407ab039cffc46f39783152770e4628d9c6c')

      avatar = Vishnu.new(openid: 'http://example.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/2a1b402420ef46577471cdc7409b0fa2c6a204db316e59ade2d805435489a067')
    end

    it 'gets url for federated openid' do
      avatar = Vishnu.new(openid: 'http://federated.com/id/user')
      expect(avatar.url).to eq('http://avatars.federated.com/avatar/50ef67971fecacf62abe7f9a002aaf6a26ff5882229a51899439dd4c7ccb9ddd')

      avatar = Vishnu.new(openid: 'http://custom-federated.com/id/user')
      expect(avatar.url).to eq('http://avatars.custom-federated.com:8080/avatar/e2014cf33d71fbf29f6976eab7f9569e7c9eae358cca0ac5b4aa536400a1c9fe')
    end

    it 'sanitizes SRV lookup result' do
      # normal, use received data
      avatar = Vishnu.new(email: 'user@nonjunk-federated.com')
      expect(avatar.url).to eq('http://hosntame.abcde.fghi.com:12345/avatar/d2b338588b00d29e3f37a5ed49244b58')

      # junk, use central service
      avatar = Vishnu.new(email: 'user@junk-federated.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/54af4ded7e369f02d8a4bcfcdc227312')

      # junk port, use central service
      avatar = Vishnu.new(email: 'user@junk2-federated.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/7f567cad8ea6f851a406c83becee21ba')

      # junk host, use central service
      avatar = Vishnu.new(email: 'user@junk3-federated.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/fa085bd0dbf5b9df0b3fd8e39eb65cd0')
    end
  end
end


spec/libravatar_spec.rb
require 'spec_helper'

RSpec.describe Libravatar do
  describe 'Gets urls' do
    before(:each) do
      dns = double('Resolv::DNS')
      allow(Resolv::DNS).to receive(:open).and_yield(dns)
      allow(dns).to receive(:getresources) do |host, type|
        expect(type).to eq(Resolv::DNS::Resource::IN::SRV)

        case host.downcase
          when '_avatars._tcp.federated.com'
            [
                Resolv::DNS::Resource::IN::SRV.new(0, 5, 80, 'avatars.federated.com'),
                Resolv::DNS::Resource::IN::SRV.new(5, 5, 80, 'not-in-priority.federated.com'),
            ]
          when '_avatars-sec._tcp.federated.com'
            [
                Resolv::DNS::Resource::IN::SRV.new(5, 5, 443, 'not-in-priority.federated.com'),
                Resolv::DNS::Resource::IN::SRV.new(0, 5, 443, 'avatars.federated.com'),
            ]
          when '_avatars._tcp.custom-federated.com'
            [Resolv::DNS::Resource::IN::SRV.new(0, 5, 8080, 'avatars.custom-federated.com')]
          when '_avatars-sec._tcp.custom-federated.com'
            [Resolv::DNS::Resource::IN::SRV.new(0, 5, 8043, 'avatars.custom-federated.com')]
          else
            []
        end
      end
    end

    it 'gets url for email' do
      # non-federated
      expect(Libravatar.new(email: 'user@example.com').url).to                                            eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')
      expect(Libravatar.new(email: 'user@example.com').to_s).to                                           eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')
      expect(Libravatar.new(email: 'USER@ExAmPlE.CoM').url).to                                            eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')
      expect(Libravatar.new(email: 'user@example.com', https: true).url).to                               eq('https://seccdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')
      expect(Libravatar.new(email: 'user@example.com', https: false).url).to                              eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af')
      expect(Libravatar.new(email: 'USER@ExAmPlE.CoM', default: 'http://example.com/avatar.png').url).to  eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af?d=http://example.com/avatar.png')
      expect(Libravatar.new(email: 'USER@ExAmPlE.CoM', size: 512, default: 'mm').url).to                  eq('http://cdn.libravatar.org/avatar/b58996c504c5638798eb6b511e6f49af?s=512&d=mm')

      # federated
      expect(Libravatar.new(email: 'user@federated.com').url).to                                          eq('http://avatars.federated.com/avatar/d69b469ded547b3ddef720a70c186322')
      expect(Libravatar.new(email: 'user@feDeRaTed.cOm', https: true).url).to                             eq('https://avatars.federated.com/avatar/d69b469ded547b3ddef720a70c186322')
      expect(Libravatar.new(email: 'USER@cuStOm-feDerated.COM').url).to                                   eq('http://avatars.custom-federated.com:8080/avatar/8df8704e4b556e0684f7c38accdaf517')
      expect(Libravatar.new(email: 'user@custom-federated.com', https: true).url).to                      eq('https://avatars.custom-federated.com:8043/avatar/8df8704e4b556e0684f7c38accdaf517')
    end

    it 'gets url for openid' do
      avatar = Libravatar.new(openid: 'http://example.com/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd')

      avatar = Libravatar.new(openid: 'https://example.com/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/82be86d8b10f6492d0eb3d6475c388044529b9d4ddf7269dec2483601b22d2e1')

      avatar = Libravatar.new(openid: 'hTTp://EXAMPLE.COM/id/Bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd')

      avatar = Libravatar.new(openid: 'hTTp://EXAMPLE.COM/id/Bob', size: 512)
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/80cd0679bb52beac4d5d388c163016dbc5d3f30c262a4f539564236ca9d49ccd?s=512')

      avatar = Libravatar.new(openid: 'http://example.com/id/bob')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/84106b275be9a7c69f3b0f77d1d504a794e6d0e4e0a068fa529d869b721f4261')

      avatar = Libravatar.new(openid: 'hTTp://EXAMPLE.COM/ID/BOB')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/10e678f0db6ead293f21fae8adb1407ab039cffc46f39783152770e4628d9c6c')

      avatar = Libravatar.new(openid: 'http://example.com')
      expect(avatar.url).to eq('http://cdn.libravatar.org/avatar/2a1b402420ef46577471cdc7409b0fa2c6a204db316e59ade2d805435489a067')

      avatar = Libravatar.new(openid: 'http://federated.com/id/user')
      expect(avatar.url).to eq('http://avatars.federated.com/avatar/50ef67971fecacf62abe7f9a002aaf6a26ff5882229a51899439dd4c7ccb9ddd')

      avatar = Libravatar.new(openid: 'http://custom-federated.com/id/user')
      expect(avatar.url).to eq('http://avatars.custom-federated.com:8080/avatar/e2014cf33d71fbf29f6976eab7f9569e7c9eae358cca0ac5b4aa536400a1c9fe')
    end
  end

  describe 'Works correctly inside' do
    it 'sanitizes openid' do
      avatar = Libravatar.new

      expect(avatar.send(:normalize_openid, 'HTTP://EXAMPLE.COM/id/Bob')).to   eq('http://example.com/id/Bob')
      expect(avatar.send(:normalize_openid, 'HTTP://EXAMPLE.COM')).to          eq('http://example.com/')
      expect(avatar.send(:normalize_openid, 'https://example.com/id/bob')).to  eq('https://example.com/id/bob')
      expect(avatar.send(:normalize_openid, 'https://eXamPlE.cOm/ID/BOB/')).to eq('https://example.com/ID/BOB/')
    end

    it 'sanitizes SRV lookup result' do
      avatar = Libravatar.new
      expect(avatar.send(:sanitize_srv_lookup, 'hosntame.abcde.fghi.com', 12345)).to    eq(['hosntame.abcde.fghi.com', 12345])
      expect(avatar.send(:sanitize_srv_lookup, 'hosntame.abcde.fghi.com', 65348283)).to eq([nil, nil])
      expect(avatar.send(:sanitize_srv_lookup, 'FNORD IMPUNTK *#(*$#&',   12345)).to    eq([nil, nil])
      expect(avatar.send(:sanitize_srv_lookup, 'FNORD IMPUNTK *#(*$#&',   65348283)).to eq([nil, nil])
    end
  end
end


lib/vishnu.rb
# frozen_string_literal: true
#
# The Vishnu class generates the avatar URL provided by the libravatar
# web service at https://www.libravatar.org
#
# Users may associate their avatar images with multiple OpenIDs and Emails.
#
# Original Author:: Kang-min Liu (http://gugod.org/)
# Fork Author:: Anton Smirnov (https://sandfox.me/)
# Copyright:: Copyright (c) 2011 Kang-min Liu
# License:: MIT
# Contributors:: https://github.com/sandfoxme/vishnu/graphs/contributors
#

require 'vishnu/version'

require 'digest/md5'
require 'digest/sha2'
require 'uri'
require 'resolv'

class Vishnu
  attr_accessor :email, :openid, :size, :default, :https

  # The options should contain :email or :openid values.  If both are
  # given, email will be used. The value of openid and email will be
  # normalized by the rule described in https://wiki.libravatar.org/api/
  #
  # List of option keys:
  #
  # - :email
  # - :openid
  # - :size An integer ranged 1 - 512, default is 80.
  # - :https Set to true to serve avatars over SSL
  # - :default URL to redirect missing avatars to, or one of these specials:
  #   "404", "mm", "identicon", "monsterid", "wavatar", "retro"
  #
  def initialize(email: nil, openid: nil, size: nil, default: nil, https: nil)
    @email   = email
    @openid  = openid
    @size    = size
    @default = default
    @https   = https
  end

  # All the values which are different between HTTP and HTTPS methods.
  PROFILES = {
    http: {
      scheme: 'http://',
      host:   'cdn.libravatar.org',
      srv:    '_avatars._tcp.',
      port:   80,
    }.freeze,
    https: {
      scheme: 'https://',
      host:   'seccdn.libravatar.org',
      srv:    '_avatars-sec._tcp.',
      port:   443,
    }.freeze
  }.freeze

  # Generate the libravatar URL
  def url
    id =
      if @email
        Digest::MD5.hexdigest(normalize_email(@email))
      else
        Digest::SHA2.hexdigest(normalize_openid(@openid))
      end

    size    = "s=#{@size}"    if @size
    default = "d=#{@default}" if @default

    # noinspection RubyScope
    # ok for them to be nil
    query = [size, default].reject(&:!).join('&')
    query = "?#{query}" unless query == ''

    base_url = get_base_url + '/avatar/'

    base_url + id + query
  end

  alias_method :to_s, :url

  private

  def profile
    PROFILES[@https ? :https : :http]
  end

  def get_target_domain
    if @email
      @email.split('@')[1]
    else
      URI.parse(@openid).host
    end
  end

  # Grab the DNS SRV records associated with the target domain,
  # and choose one according to RFC2782.
  def srv_lookup
    Resolv::DNS.open do |dns|
      resources = dns.getresources(
        profile[:srv] + get_target_domain,
        Resolv::DNS::Resource::IN::SRV
      ).to_a

      return [nil, nil] unless resources.any?

      min_priority = resources.map(&:priority).min
      resources.delete_if { |r| r.priority != min_priority }

      weight_sum = resources.inject(0) { |sum, r| sum + r.weight }.to_f

      resource = resources.max_by do |r|
        if r.weight == 0
          0
        else
          rand**(weight_sum / r.weight)
        end
      end

      return sanitize_srv_lookup(resource.target.to_s, resource.port)
    end
  end

  def get_base_url
    target, port = srv_lookup

    if target && port
      port_fragment = port != profile[:port] ? ':' + port.to_s : ''
      profile[:scheme] + target.to_s + port_fragment
    else
      profile[:scheme] + profile[:host]
    end
  end

  def sanitize_srv_lookup(hostname, port)
    unless hostname.match(/^[0-9a-zA-Z\-.]+$/) && 1 <= port && port <= 65_535
      return [nil, nil]
    end

    [hostname, port]
  end

  def normalize_email(email)
    email.downcase
  end

  # Normalize an openid URL following the description on libravatar.org
  def normalize_openid(url)
    parsed_url = URI.parse(url)
    parsed_url.host   = parsed_url.host.downcase
    parsed_url.scheme = parsed_url.scheme.downcase
    if parsed_url.path == '' && parsed_url.fragment.nil?
      parsed_url.path = '/'
    end

    parsed_url.to_s
  end
end
